ctpolicy/loglist/loglist.go

```diff
@@ -0,0 +1,317 @@
+		}
+	}
+
+	if len(remaining) > 0 {
+		missed := make([]string, len(remaining))
+		for name := range remaining {
+			missed = append(missed, fmt.Sprintf("%q", name))
+		}
+		return nil, fmt.Errorf("failed to find logs matching name(s): %s", strings.Join(missed, ", "))
+	}
+
+	return newList, nil
+}
+
+// forPurpose returns a new log list containing only those logs whose states are
+// acceptable for the given purpose. It returns an error if the purpose is
+// Issuance or Validation and the set of remaining logs is too small to satisfy
+// the Google "two operators" log policy.
+func (ll List) forPurpose(p purpose) (List, error) {
+	newList := make(List)
+	for operator, group := range ll {
+		newGroup := make(OperatorGroup)
+		for id, log := range group {
+			if !usableForPurpose(log.State, p) {
+				continue
+			}
+
+			newLog := Log{
+				Name:           log.Name,
+				Url:            log.Url,
+				Key:            log.Key,
+				State:          log.State,
+				StartInclusive: log.StartInclusive,
+				EndExclusive:   log.EndExclusive,
+			}
+
+			newGroup[id] = newLog
+		}
+		if len(newGroup) > 0 {
+			newList[operator] = newGroup
+		}
+	}
+
+	if len(newList) < 2 && p != Informational {
+		return nil, errors.New("log list does not have enough groups to satisfy Chrome policy")
+	}
+
+	return newList, nil
+}
+
+// OperatorForLogID returns the Name of the Group containing the Log with the
+// given ID, or an error if no such log/group can be found.
+func (ll List) OperatorForLogID(logID string) (string, error) {
+	for op, group := range ll {
+		if _, found := group[logID]; found {
+			return op, nil
+		}
+	}
+	return "", fmt.Errorf("no log with ID %q found", logID)
+}
+
+// Permute returns the list of operator group names in a randomized order.
+func (ll List) Permute() []string {
+	keys := make([]string, 0, len(ll))
+	for k := range ll {
+		keys = append(keys, k)
+	}
+
+	result := make([]string, len(ll))
+	for i, j := range rand.Perm(len(ll)) {
+		result[i] = keys[j]
+	}
+
+	return result
+}
+
+// PickOne returns the URI and Public Key of a single randomly-selected log
+// which is run by the given operator and whose temporal interval includes the
+// given expiry time. It returns an error if no such log can be found.
+func (ll List) PickOne(operator string, expiry time.Time) (string, string, error) {
+	group, ok := ll[operator]
+	if !ok {
+		return "", "", fmt.Errorf("no log operator group named %q", operator)
+	}
+
+	candidates := make([]Log, 0)
+	for _, log := range group {
+		if log.StartInclusive.IsZero() || log.EndExclusive.IsZero() {
+			candidates = append(candidates, log)
+			continue
+		}
+
+		if (log.StartInclusive.Equal(expiry) || log.StartInclusive.Before(expiry)) && log.EndExclusive.After(expiry) {
+			candidates = append(candidates, log)
+		}
+	}
+
+	if len(candidates) < 1 {
+		return "", "", fmt.Errorf("no log found for group %q and expiry %s", operator, expiry)
+	}
+
+	log := candidates[rand.Intn(len(candidates))]
```

Comment:
Nit, given that calling `rand.Intn()` with an integer `< 1` will result in a panic it might make sense to connect these two with an `else` statement, a comment stating as such, or both.

Reply:
Waiting for a language that has not just signed and unsigned int types, but positive/negative/non-negative/non-zero int types.

Reply:
[clears throat](https://doc.rust-lang.org/nightly/std/num/struct.NonZeroU64.html)

Reply:
You two just made my day 😅